src/modm/platform/can/atsam-fdcan/can.hpp.in

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright (c) 2019, 2023, Raphael Lehmann
+ * Copyright (c) 2021, Christopher Durand
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <optional>
+
+#include <modm/architecture/interface/can.hpp>
+#include <modm/architecture/interface/can_filter.hpp>
+#include <modm/platform/gpio/connector.hpp>
+#include "../device.hpp"
+
+#include "can_bit_timings.hpp"
+
+#define SRAMCAN_BASE (0) // TODO
```

Comment:
There is an important difference regarding the message RAM compared to STM32G4. On those controllers the message RAM is a dedicated hardware SRAM exclusively reserved for the CAN peripheral.

On SAMx7x there is no such thing. The data structure resides in normal RAM and the addresses for the start of the RX and TX sections are configured in the `MCAN_TXBC` and `MCAN_RXBC` registers. The user has to allocate adequately sized 4-byte aligned blocks of memory and write their addresses to the respective registers. The size of buffers and filter counts are also fully customizable.

As a start you could just assume the fixed layout of the STM32G4 version and defined an equally sized array as the message RAM.

Reply:
That is what they are doing in the vendor HAL: https://github.com/osmocom/atmel-asf-projects/blob/3634cc8e5d3124e3cf6cf46c0f1d5385ff511426/sam/drivers/mcan/mcan.c#L65